In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline

We've talked about Random Forests. Now it's time to build one.

Here we'll use data from Lending Club to predict the state of a loan given some information about it. You can find the dataset [here](https://www.lendingclub.com/info/download-data.action). We'll use 2015 data. ([Thinkful mirror](https://www.dropbox.com/s/m7z42lubaiory33/LoanStats3d.csv?dl=0))

In [2]:
# Replace the path with the correct path for your data.
y2015 = pd.read_csv(
    'https://www.dropbox.com/s/0so14yudedjmm5m/LoanStats3d.csv?dl=1',
    skipinitialspace=True,
    header=1
)

# Note the warning about dtypes.

C:\Users\asus\Anaconda3\lib\site-packages\IPython\core\interactiveshell.py:2785: DtypeWarning: Columns (0,19,55) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


In [3]:
y2015.head()

id   member_id  loan_amnt  funded_amnt  funded_amnt_inv        term  \
0  68009401  72868139.0    16000.0      16000.0          16000.0   60 months   
1  68354783  73244544.0     9600.0       9600.0           9600.0   36 months   
2  68466916  73356753.0    25000.0      25000.0          25000.0   36 months   
3  68466961  73356799.0    28000.0      28000.0          28000.0   36 months   
4  68495092  73384866.0     8650.0       8650.0           8650.0   36 months   

  int_rate  installment grade sub_grade            ...              \
0   14.85%       379.39     C        C5            ...               
1    7.49%       298.58     A        A4            ...               
2    7.49%       777.55     A        A4            ...               
3    6.49%       858.05     A        A2            ...               
4   19.89%       320.99     E        E3            ...               

  num_tl_90g_dpd_24m num_tl_op_past_12m pct_tl_nvr_dlq  percent_bc_gt_75  \
0                0.0                2.0           78.9               0.0   
1                0.0                2.0          100.0              66.7   
2                0.0                0.0          100.0              20.0   
3                0.0                0.0           91.7              22.2   
4                0.0               12.0          100.0              50.0   

  pub_rec_bankruptcies tax_liens tot_hi_cred_lim total_bal_ex_mort  \
0                  0.0       2.0        298100.0           31329.0   
1                  0.0       0.0         88635.0           55387.0   
2                  0.0       0.0        373572.0           68056.0   
3                  0.0       0.0        304003.0           74920.0   
4                  1.0       0.0         38998.0           18926.0   

  total_bc_limit total_il_high_credit_limit  
0       281300.0                    13400.0  
1        12500.0                    75635.0  
2        38400.0                    82117.0  
3        41500.0                    42503.0  
4         2750.0                    18248.0  

[5 rows x 111 columns]

## The Blind Approach

Now, as we've seen before, creating a model is the easy part. Let's try just using everything we've got and throwing it without much thought into a Random Forest. SKLearn requires the independent variables to be be numeric, and all we want is dummy variables so let's use `get_dummies` from Pandas to generate a dummy variable for every categorical colummn and see what happens off of this kind of naive approach.

In [4]:
from sklearn import ensemble
from sklearn.model_selection import cross_val_score

rfc = ensemble.RandomForestClassifier()
X = y2015.drop('loan_status', 1)
Y = y2015['loan_status']
X = pd.get_dummies(X)

cross_val_score(rfc, X, Y, cv=5)

MemoryError: 

Did your kernel die? My kernel died.

Guess it isn't always going to be that easy...

Can you think of what went wrong?

(You're going to have to reset your kernel and reload the column, BUT DON'T RUN THE MODEL AGAIN OR YOU'LL CRASH THE KERNEL AGAIN!)

## Data Cleaning

Well, `get_dummies` can be a very memory intensive thing, particularly if data are typed poorly. We got a warning about that earlier. Mixed data types get converted to objects, and that could create huge problems. Our dataset is about 400,000 rows. If there's a bad type there its going to see 400,000 distinct values and try to create dummies for all of them. That's bad. Lets look at all our categorical variables and see how many distinct counts there are...

In [5]:
categorical = y2015.select_dtypes(include=['object'])
for i in categorical:
    column = categorical[i]
    print(i)
    print(column.nunique())

id
421097
term
2
int_rate
110
grade
7
sub_grade
35
emp_title
120812
emp_length
11
home_ownership
4
verification_status
3
issue_d
12
loan_status
7
pymnt_plan
1
url
421095
desc
34
purpose
14
title
27
zip_code
914
addr_state
49
earliest_cr_line
668
revol_util
1211
initial_list_status
2
last_pymnt_d
25
next_pymnt_d
4
last_credit_pull_d
26
application_type
2
verification_status_joint
3


Well that right there is what's called a problem. Some of these have over a hundred thousand distinct types. Lets drop the ones with over 30 unique values, converting to numeric where it makes sense. In doing this there's a lot of code that gets written to just see if the numeric conversion makes sense. It's a manual process that we'll abstract away and just include the conversion.

You could extract numeric features from the dates, but here we'll just drop them. There's a lot of data, it shouldn't be a huge problem.

In [6]:
# Convert ID and Interest Rate to numeric.
y2015['id'] = pd.to_numeric(y2015['id'], errors='coerce')
y2015['int_rate'] = pd.to_numeric(y2015['int_rate'].str.strip('%'), errors='coerce')

# Drop other columns with many unique variables
y2015.drop(['url', 'emp_title', 'zip_code', 'earliest_cr_line', 'revol_util',
            'sub_grade', 'addr_state', 'desc'], 1, inplace=True)

Wonder what was causing the dtype error on the id column, which _should_ have all been integers? Let's look at the end of the file.

In [7]:
y2015.tail()

id   member_id  loan_amnt  funded_amnt  funded_amnt_inv  \
421092  36271333.0  38982739.0    13000.0      13000.0          13000.0   
421093  36490806.0  39222577.0    12000.0      12000.0          12000.0   
421094  36271262.0  38982659.0    20000.0      20000.0          20000.0   
421095         NaN         NaN        NaN          NaN              NaN   
421096         NaN         NaN        NaN          NaN              NaN   

              term  int_rate  installment grade emp_length  \
421092   60 months     15.99       316.07     D    5 years   
421093   60 months     19.99       317.86     E     1 year   
421094   36 months     11.99       664.20     B  10+ years   
421095         NaN       NaN          NaN   NaN        NaN   
421096         NaN       NaN          NaN   NaN        NaN   

                   ...             num_tl_90g_dpd_24m  num_tl_op_past_12m  \
421092             ...                            0.0                 3.0   
421093             ...                            1.0                 2.0   
421094             ...                            0.0                 1.0   
421095             ...                            NaN                 NaN   
421096             ...                            NaN                 NaN   

       pct_tl_nvr_dlq percent_bc_gt_75 pub_rec_bankruptcies tax_liens  \
421092          100.0             50.0                  1.0       0.0   
421093           95.0             66.7                  0.0       0.0   
421094          100.0             50.0                  0.0       1.0   
421095            NaN              NaN                  NaN       NaN   
421096            NaN              NaN                  NaN       NaN   

       tot_hi_cred_lim total_bal_ex_mort  total_bc_limit  \
421092         51239.0           34178.0         10600.0   
421093         96919.0           58418.0          9700.0   
421094         43740.0           33307.0         41700.0   
421095             NaN               NaN             NaN   
421096             NaN               NaN             NaN   

        total_il_high_credit_limit  
421092                     33239.0  
421093                     69919.0  
421094                         0.0  
421095                         NaN  
421096                         NaN  

[5 rows x 103 columns]

In [8]:
# Remove two summary rows at the end that don't actually contain data.
y2015 = y2015[:-2]

Now this should be better. Let's try again.

In [9]:
pd.get_dummies(y2015)

id   member_id  loan_amnt  funded_amnt  funded_amnt_inv  \
0       68009401.0  72868139.0    16000.0      16000.0          16000.0   
1       68354783.0  73244544.0     9600.0       9600.0           9600.0   
2       68466916.0  73356753.0    25000.0      25000.0          25000.0   
3       68466961.0  73356799.0    28000.0      28000.0          28000.0   
4       68495092.0  73384866.0     8650.0       8650.0           8650.0   
5       68506798.0  73396623.0    23000.0      23000.0          23000.0   
6       68566886.0  73456723.0    29900.0      29900.0          29900.0   
7       68577849.0  73467703.0    18000.0      18000.0          18000.0   
8       66310712.0  71035433.0    35000.0      35000.0          35000.0   
9       68476807.0  73366655.0    10400.0      10400.0          10400.0   
10      68341763.0  72928789.0    20000.0      20000.0          20000.0   
11      68356421.0  73246236.0    22400.0      22400.0          22400.0   
12      68426545.0  73316365.0    16000.0      16000.0          16000.0   
13      68407277.0  73297111.0     3600.0       3600.0           3600.0   
14      68355089.0  73244858.0    24700.0      24700.0          24700.0   
15      68466926.0  73356763.0    10000.0      10000.0          10000.0   
16      68338832.0  73542831.0     1400.0       1400.0           1400.0   
17      68426831.0  73316657.0    11950.0      11950.0          11950.0   
18      66624733.0  71350547.0    18000.0      18000.0          18000.0   
19      68616873.0  73519699.0     8000.0       8000.0           8000.0   
20      67275481.0  72077199.0    20000.0      20000.0          20000.0   
21      68476668.0  73366513.0    20000.0      20000.0          20000.0   
22      68476702.0  73366548.0    28000.0      28000.0          28000.0   
23      68436666.0  73326493.0     5000.0       5000.0           5000.0   
24      68596180.0  73485987.0    20000.0      20000.0          20000.0   
25      67849662.0  72708407.0     4225.0       4225.0           4225.0   
26      68526883.0  73416719.0    15000.0      15000.0          15000.0   
27      68607141.0  73509983.0    17600.0      17600.0          17600.0   
28      68476715.0  73366561.0     6000.0       6000.0           6000.0   
29      67715283.0  72569997.0    16000.0      16000.0          16000.0   
...            ...         ...        ...          ...              ...   
421065  36631385.0  39374138.0    13000.0      13000.0          13000.0   
421066  36231083.0  38942474.0    15000.0      15000.0          15000.0   
421067  36691397.0  39434133.0    20000.0      20000.0          20000.0   
421068  36743377.0  39486112.0     4200.0       4200.0           4200.0   
421069  36331554.0  39042967.0    26950.0      26950.0          26950.0   
421070  36460697.0  39182359.0    23000.0      23000.0          23000.0   
421071  35773415.0  38468842.0    18700.0      18700.0          18700.0   
421072  36250801.0  38962184.0     6000.0       6000.0           6000.0   
421073  36371802.0  39103231.0    15000.0      15000.0          15000.0   
421074  36440728.0  39152101.0    25000.0      25000.0          25000.0   
421075  36361815.0  39093260.0    26500.0      26500.0          26500.0   
421076  35591131.0  38288343.0    25000.0      25000.0          25000.0   
421077  36201871.0  38913321.0    21000.0      21000.0          21000.0   
421078  36381756.0  39013193.0     8000.0       8000.0           8000.0   
421079  36231724.0  38943173.0    12000.0      12000.0          12000.0   
421080  36301729.0  39023153.0    10000.0      10000.0          10000.0   
421081  36250208.0  38961532.0    13150.0      13150.0          13150.0   
421082  36241316.0  38952731.0     6225.0       6225.0           6225.0   
421083  36231718.0  38943165.0    10775.0      10775.0          10775.0   
421084  36381704.0  39013139.0     7000.0       7000.0           7000.0   
421085  36421485.0  39142898.0     4000.0       4000.0           4000.0   
421086  35094480.0  37777829.0     7500.0

It finally works! We had to sacrifice sub grade, state address and description, but that's fine. If you want to include them you could run the dummies independently and then append them back to the dataframe.

## Second Attempt

Now let's try this model again.

We're also going to drop NA columns, rather than impute, because our data is rich enough that we can probably get away with it.

This model may take a few minutes to run.

In [10]:
from sklearn import ensemble
from sklearn.model_selection import cross_val_score

rfc = ensemble.RandomForestClassifier()
X = y2015.drop('loan_status', 1)
Y = y2015['loan_status']
X = pd.get_dummies(X)
X = X.dropna(axis=1)

cross_val_score(rfc, X, Y, cv=10)

array([0.98012396, 0.98064639, 0.98152502, 0.98185747, 0.97442413,
       0.97611019, 0.96931772, 0.98038424, 0.97993208, 0.97855413])

The score cross validation reports is the accuracy of the tree. Here we're about 98% accurate.

That works pretty well, but there are a few potential problems. Firstly, we didn't really do much in the way of feature selection or model refinement. As such there are a lot of features in there that we don't really need. Some of them are actually quite impressively useless.

There's also some variance in the scores. The fact that one gave us only 93% accuracy while others gave higher than 98 is concerning. This variance could be corrected by increasing the number of estimators. That will make it take even longer to run, however, and it is already quite slow.

## DRILL: Third Attempt

So here's your task. Get rid of as much data as possible without dropping below an average of 90% accuracy in a 10-fold cross validation.

You'll want to do a few things in this process. First, dive into the data that we have and see which features are most important. This can be the raw features or the generated dummies. You may want to use PCA or correlation matrices.

Can you do it without using anything related to payment amount or outstanding principal? How do you know?

In [38]:
y2015_loan_status = y2015
y2015_loan_status = pd.get_dummies(y2015_loan_status, columns=['loan_status'], dtype='float64')

In [40]:
# Find the most correlated variables to loan status.
pd.set_option('display.max_columns', 103)
pd.set_option('display.max_rows', 103)
y2015_loan_status.corr()

id  member_id  loan_amnt  funded_amnt  \
id                              1.000000   0.997540  -0.008288    -0.008288   
member_id                       0.997540   1.000000  -0.008620    -0.008620   
loan_amnt                      -0.008288  -0.008620   1.000000     1.000000   
funded_amnt                    -0.008288  -0.008620   1.000000     1.000000   
funded_amnt_inv                -0.008554  -0.008885   0.999994     0.999994   
int_rate                       -0.053402  -0.053447   0.140572     0.140572   
installment                    -0.012919  -0.013274   0.941205     0.941205   
annual_inc                      0.015055   0.014958   0.305734     0.305734   
dti                             0.005692   0.005806   0.006103     0.006103   
delinq_2yrs                     0.002300   0.002261  -0.010424    -0.010424   
inq_last_6mths                  0.000660   0.000715  -0.039040    -0.039040   
mths_since_last_delinq          0.006322   0.006186  -0.034863    -0.034863   
mths_since_last_record         -0.037319  -0.037366  -0.014108    -0.014108   
open_acc                        0.006309   0.006111   0.188214     0.188214   
pub_rec                         0.003501   0.003635  -0.088899    -0.088899   
revol_bal                       0.003774   0.003745   0.334656     0.334656   
total_acc                      -0.014479  -0.014673   0.208430     0.208430   
out_prncp                       0.192187   0.191883   0.623926     0.623926   
out_prncp_inv                   0.192058   0.191755   0.623961     0.623961   
total_pymnt                    -0.219883  -0.219945   0.706184     0.706184   
total_pymnt_inv                -0.220078  -0.220139   0.706157     0.706157   
total_rec_prncp                -0.188353  -0.188443   0.556266     0.556266   
total_rec_int                  -0.165885  -0.165861   0.736585     0.736585   
total_rec_late_fee             -0.015779  -0.015748   0.051888     0.051888   
recoveries                     -0.070027  -0.069799   0.077870     0.077870   
collection_recovery_fee        -0.070594  -0.070372   0.079190     0.079190   
last_pymnt_amnt                -0.068844  -0.068974   0.259864     0.259864   
collections_12_mths_ex_med      0.011953   0.011891  -0.027454    -0.027454   
mths_since_last_major_derog     0.021221   0.021181  -0.018556    -0.018556   
policy_code                          NaN        NaN        NaN          NaN   
annual_inc_joint                0.056260   0.056358   0.438866     0.438866   
dti_joint                       0.092846   0.093700   0.018769     0.018769   
acc_now_delinq                 -0.004834  -0.004732   0.000645     0.000645   
tot_coll_amt                    0.002312   0.002329  -0.021437    -0.021437   
tot_cur_bal                     0.011241   0.010903   0.325352     0.325352   
open_acc_6m                     0.012746   0.008560  -0.040487    -0.040487   
open_il_6m                      0.002460   0.000563   0.027938     0.027938   
open_il_12m                     0.006383   0.005267  -0.013290    -0.013290   
open_il_24m                     0.001371  -0.002238   0.024537     0.024537   
mths_since_rcnt_il              0.002310   0.003673  -0.010985    -0.010985   
total_bal_il                    0.009510   0.005955   0.135564     0.135564   
il_util                         0.009614   0.008941  -0.107324    -0.107324   
open_rv_12m                     0.011797   0.006729  -0.051313    -0.051313   
open_rv_24m                     0.004960   0.003529  -0.028005    -0.028005   
max_bal_bc                     -0.000543  -0.003295   0.412165     0.412165   
all_util                        0.000399   0.000448   0.019117     0.019117   
total_rev_hi_lim                0.019249   0.019148   0.330877     0.330877   
inq_fi                         -0.013579  -0.013022   0.009638     0.009638   
total_cu_tl                    -0.000914  -0.005038   0.071500     0.071500   
inq_last_12m                   -0.001412  -0.001735   0.007574     0.007574   
acc_open_

In [41]:
# Select most useful features: those above 0.50 correlation to loan statuses
X = y2015[['out_prncp', 'out_prncp_inv', 'total_pymnt', 'total_pymnt_inv', 'total_rec_prncp', 'last_pymnt_amnt']]
cross_val_score(rfc, X, Y, cv=10)

array([0.79641899, 0.95901308, 0.94303151, 0.95359882, 0.9501781 ,
       0.94072667, 0.90902182, 0.92574034, 0.94518726, 0.95953071])

Decreasing my columns from 103 down to 6, I still managed to get accuracies between 91 - 96% accuracies, with one outlier of 80%.

Can I get good accuracies still without anything related to payment amount or outstanding principal?

In [48]:
pd.set_option('display.max_info_columns', 103)
y2015

id   member_id  loan_amnt  funded_amnt  funded_amnt_inv  \
0       68009401.0  72868139.0    16000.0      16000.0          16000.0   
1       68354783.0  73244544.0     9600.0       9600.0           9600.0   
2       68466916.0  73356753.0    25000.0      25000.0          25000.0   
3       68466961.0  73356799.0    28000.0      28000.0          28000.0   
4       68495092.0  73384866.0     8650.0       8650.0           8650.0   
5       68506798.0  73396623.0    23000.0      23000.0          23000.0   
6       68566886.0  73456723.0    29900.0      29900.0          29900.0   
7       68577849.0  73467703.0    18000.0      18000.0          18000.0   
8       66310712.0  71035433.0    35000.0      35000.0          35000.0   
9       68476807.0  73366655.0    10400.0      10400.0          10400.0   
10      68341763.0  72928789.0    20000.0      20000.0          20000.0   
11      68356421.0  73246236.0    22400.0      22400.0          22400.0   
12      68426545.0  73316365.0    16000.0      16000.0          16000.0   
13      68407277.0  73297111.0     3600.0       3600.0           3600.0   
14      68355089.0  73244858.0    24700.0      24700.0          24700.0   
15      68466926.0  73356763.0    10000.0      10000.0          10000.0   
16      68338832.0  73542831.0     1400.0       1400.0           1400.0   
17      68426831.0  73316657.0    11950.0      11950.0          11950.0   
18      66624733.0  71350547.0    18000.0      18000.0          18000.0   
19      68616873.0  73519699.0     8000.0       8000.0           8000.0   
20      67275481.0  72077199.0    20000.0      20000.0          20000.0   
21      68476668.0  73366513.0    20000.0      20000.0          20000.0   
22      68476702.0  73366548.0    28000.0      28000.0          28000.0   
23      68436666.0  73326493.0     5000.0       5000.0           5000.0   
24      68596180.0  73485987.0    20000.0      20000.0          20000.0   
25      67849662.0  72708407.0     4225.0       4225.0           4225.0   
26      68526883.0  73416719.0    15000.0      15000.0          15000.0   
27      68607141.0  73509983.0    17600.0      17600.0          17600.0   
28      68476715.0  73366561.0     6000.0       6000.0           6000.0   
29      67715283.0  72569997.0    16000.0      16000.0          16000.0   
30      68526907.0  73416744.0    16000.0      16000.0          16000.0   
31      68533595.0  73423325.0    12000.0      12000.0          12000.0   
32      68415473.0  73305264.0    25000.0      25000.0          25000.0   
33      68341789.0  73544827.0    24250.0      24250.0          24250.0   
34      68367011.0  73256863.0    21000.0      21000.0          21000.0   
35      68356614.0  73246437.0     7200.0       7200.0           7200.0   
36      68446769.0  73336592.0     7000.0       7000.0           7000.0   
37      68407301.0  73297138.0    27500.0      27500.0          27500.0   
38      68537655.0  73427491.0    16800.0      16800.0          16800.0   
39      68615044.0  73517819.0    16000.0      16000.0          16000.0   
40      68426865.0  73316695.0    20000.0      20000.0          20000.0   
41      68547583.0  73437441.0     8650.0       8650.0           8650.0   
42      68566951.0  73456790.0    10000.0      10000.0          10000.0   
43      68527009.0  73416854.0    20000.0      20000.0          20000.0   
44      68476676.0  73366521.0    20000.0      20000.0          20000.0   
45      66796130.0  71585842.0     8800.0       8800.0           8800.0   
46      68387134.0  73276991.0     2500.0       2500.0           2500.0   
47      68617034.0  73519870.0    14650.0      14650.0          14650.0   
48      68446771.0  73336594.0     7200.0       7200.0           7200.0   
49      68566925.0  73456763.0    11000.0      11000.0          11000.0   
50      68587709.0  73477552.0    21000.0      21000.0          21000.0   
...            ...         ...        ...          ...              ...   
421044  36799561.0  39552273.0    17025.0

In [56]:
X = y2015.drop(columns=['loan_status', 'total_pymnt', 'total_pymnt_inv', 'total_rec_int', 'last_pymnt_amnt', 'out_prncp'])

In [57]:
X = pd.get_dummies(X)
X = X.dropna(axis=1)

cross_val_score(rfc, X, Y, cv=10)

array([0.97691814, 0.97432975, 0.97941156, 0.97706063, 0.96079316,
       0.97295179, 0.88313662, 0.97779572, 0.97608474, 0.97522918])

Yes; if I keep the remainder of the 102 columns, I can still have roughly 96-97% accuracy (one outlier of 88%), even without variables related to payment or outstanding balance.